### PR TITLE
Add idempotent primary keys on tables missing them

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR adds primary keys on the missing tables in idempotent way

Tables updated are:
- cloud.op_user_stats_log
- cloud_usage.usage_ip_address
- cloud_usage.usage_load_balancer_policy
- cloud_usage.usage_network_offering
- cloud_usage.usage_port_forwarding
- cloud_usage.usage_security_group
- cloud_usage.usage_snapshot_on_primary
- cloud_usage.usage_storage
- cloud_usage.usage_vm_instance
- cloud_usage.usage_vmsnapshot
- cloud_usage.usage_volume
- cloud_usage.usage_vpn_user

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by GabrielBrascher — COMMENTED** [review]
Thanks for the PR @nvazquez!
Code LGTM, but I did not test it. Seems simple but yet critical enough to require some elaborated tests.
How do you advise testing this @nvazquez?

**Comment by nvazquez** [comment]
Thanks @GabrielBrascher!

To test the DB additions:
 - We should be able to run the upgrade path script multiple times and still get the same results without any failure.
 - Verify that each table from the list contains a primary key

The java code simply reflects the new DB changes, however, the modifications can impact the usage server (the only table modified on the `cloud` schema simply adds a new column, but there are many tables in the `cloud_usage` schema which we have modified an existing column and an existing unique index).
- To test it I would suggest checking the usage serve keeps working as usual
- Check persisted usage data after the upgrade is consistent with the data prior to the upgrade

File: engine/schema/src/main/java/com/cloud/usage/UsageVMSnapshotVO.java [hunk 1]
```diff
@@ -30,9 +33,14 @@
 @Table(name = "usage_vmsnapshot")
 public class UsageVMSnapshotVO implements InternalIdentity {
 
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     @Column(name = "id")
+    private Long id;
+
+    @Column(name = "volume_id")
     // volumeId
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/usage/UsageVMSnapshotVO.java:42]: let's get rid of this silly comment

File: engine/schema/src/main/java/com/cloud/usage/UsageSnapshotOnPrimaryVO.java [hunk 2]
```diff
@@ -30,9 +33,14 @@
 @Table(name = "usage_snapshot_on_primary")
 public class UsageSnapshotOnPrimaryVO implements InternalIdentity {
 
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     @Column(name = "id")
+    private Long id;
+
+    @Column(name = "volume_id")
     // volumeId
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/usage/UsageSnapshotOnPrimaryVO.java:42]: let's get rid of this silly comment

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@sureshanaparti this is a fix, but does contain db changes. i think it should go in the oldest possible release but their are upgrade risks to it. At your discretion!

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by GabrielBrascher — APPROVED** [review]
Thanks for the PR @nvazquez.
Code LGTM

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: engine/schema/src/main/resources/META-INF/db/schema-41600to41610.sql [hunk 3]
```diff
@@ -17,4 +17,111 @@
...
+
+-- Idempotent ADD COLUMN
+DROP PROCEDURE IF EXISTS `cloud`.`IDEMPOTENT_ADD_COLUMN`;
+CREATE PROCEDURE `cloud`.`IDEMPOTENT_ADD_COLUMN` (
+    IN in_table_name VARCHAR(200)
+, IN in_column_name VARCHAR(200)
+, IN in_column_definition VARCHAR(1000)
+)
+BEGIN
+    DECLARE CONTINUE HANDLER FOR 1060 BEGIN END; SET @ddl = CONCAT('ALTER TABLE ', in_table_name); SET @ddl = CONCAT(@ddl, ' ', 'ADD COLUMN') ; SET @ddl = CONCAT(@ddl, ' ', in_column_name); SET @ddl = CONCAT(@ddl, ' ', in_column_definition); PREPARE stmt FROM @ddl; EXECUTE stmt; DEALLOCATE PREPARE stmt; END;
```
Comment (GutoVeronezi) [inline engine/schema/src/main/resources/META-INF/db/schema-41600to41610.sql:32]: I think we could separate the commands in multiple lines, to improve readability. If this suggestion get accepted, we should also apply to the lines below.

**Review by rohityadavcloud — APPROVED** [review]
LGTM, needs testing cc @borisstoyanov @vladimirpetrov

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@nvazquez can you fix the conflicts.

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Pull request merged**